Framework/PythonInterface/plugins/algorithms/AlignAndFocusPowderFromFiles.py

```diff
@@ -439,6 +405,158 @@ def PyExec(self):
+        If no cache found for any partial sum, return the filename list.
+        If a cache found for a partial sum, continue to work on the remained files by calling this function.
+        When a cache is found, it will be loaded and accumulated, and then deleted.
+        """
+        N = len(filenames)
+        found = False
+        for length in range(N, 1, -1):
+            for start in range(N-length+1):
+                end = start+length-1
+                files1 = filenames[start:end+1]
+                summed_cache_file = self.__get_grp_cache_fn(files1)
+                if os.path.exists(summed_cache_file):
+                    self.log().notice('Found cache for {}'.format(files1))
+                    wkspname = self.__get_grp_ws_name(files1)
+                    self.log().notice('Loading cache from {}'.format(summed_cache_file))
+                    LoadNexusProcessed(Filename=summed_cache_file, OutputWorkspace=wkspname)
+                    self.__accumulate(wkspname, finalname, unfocusname, unfocusname_file)
+                    found = True
+                    break
+            if found:
+                break
+            continue
+        if not found:
+            return filenames
+        remained = filenames[:start] + filenames[end+1:]
+        if remained:
+            return self.__find_caches(remained, finalname, unfocusname, unfocusname_file, firstcall=False)
+        return []
+
+    def __get_grp_ws_name(self, group):
+        def _(filename):
+            return os.path.split(filename)[-1].split('.')[0]
+        return _(group[0]) + "-" + _(group[-1])
+
+    def __get_grp_cache_fn(self, group):
+        wsname = self.__get_grp_ws_name(group)
+        filenames_str = ','.join(group)
+        newprop = 'files_to_sum={}'.format(filenames_str)
+        return self.__getCacheName('summed_'+wsname, additional_props=[newprop])
+
+    def __processFiles(self, files, useCaching, unfocusname, unfocusname_file):
+        """process given files, separate them to "grains". Sum each grain, and add the grain sum to final sum
+        """
+        finalname = self.getPropertyValue('OutputWorkspace')
+        N = len(files)
+        import math
+        if N>3: #9
+            n = int(math.sqrt(N)) # grain size
+        else:
+            n = N
+        for (i, f) in enumerate(files):
+            wkspname = self.__processFile2_withcache(f, useCaching, unfocusname, unfocusname_file)
+            # accumulate into partial sum
+            grain_start = i//n*n
+            grain_end = min((i//n+1)*n, N)
+            grain = files[grain_start:grain_end]
+            partialsum_wkspname = self.__get_grp_ws_name(grain)
+            self.__accumulate(
+                wkspname, partialsum_wkspname, unfocusname, unfocusname_file)
+            if i==grain_end-1:
+                if useCaching and len(grain)>1:
+                    # save partial cache
+                    self.__saveSummedGroupToCache(grain)
```

Comment:
Don't you need to either delete the "grain" or reset it?

Reply:
Each "grain" is named differently. So no reset is needed. And when it is "accumulated" to the final sum, it is deleted